README.md
# Deducto

Deducto is a learning activity to help children learn the art of
inductive logic through pattern recognition.


## Inspiration

The initial code of Deducto was written in Perl by Jon Orwant, who
wrote it as part of this doctorial thesis on EGGG, a game generation
language, under the guidance of Walter Bender.

## Authors

Walter wrote a Javascript version for the web. This version was written by 
scratch by Walter in order to better match the standard Sugar look and feel.

The Deducto Activity was written by a team of engineers at SEETA
(http://seeta.in): Ashita Dadlani, Anisha Arora, Satyajeet Singh, and
Manusheel Gupta

A Glade version of Deducto was developed by Assim Deodia, Vijay
Majumadar and Satyajeet Singh under the guidance of Manusheel Gupta
and Walter Bender.

A Non-sugarized pygtk version of Deducto was developed by Naveen
Aggarwal and Pankhuri Gupta under the guidance of Manusheel Gupta and
Vijay Majumadar.






NEWS
NEWS

12
* Fix typo preventing game from launching (Walter)

11
* L10n refresh (Chris Leonard et al)
* Use markdown for readme (Emily Ong)
* Ported to gtk3 + pep8 fixes (Ignacio Rodriguez)
* Fix PyGI warnings (Rahul Bothra)
* Port Telepathy Static Binding to TelepathyGLib (Rahul Bothra)
* Port to Python 3 (Ibiam Chihurumnaya)
* translate.sugarlabs.org synchronisation (Ibiam Chihurumnaya)
* Regenerate POT file (Ibiam Chihurumnaya)

10

* New translations
* Added repository to activity.info

9

ENHANCEMENT:

* Ignacio Rodríguez port to gtk3

8

ENHANCEMENT:

* New translations

7

BUG FIX:

* Initialize _python_code to None in case where loading

  levels is aborted (Alan Jhonn Aguiar Schwyn)

6

ENHANCEMENT:

* Sharing enabled

5

ENHANCEMENT:

* Rewritten from scratch to be more Sugar-like.


game.py
# -*- coding: utf-8 -*-
# Copyright (c) 2012 Walter Bender
# Ported to Gtk3:
# Ignacio Rodríguez <ignaciorodriguez@sugarlabs.org>
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# You should have received a copy of the GNU General Public License
# along with this library; if not, write to the Free Software
# Foundation, 51 Franklin Street, Suite 500 Boston, MA 02110-1335 USA

from gi.repository import Gdk, GdkPixbuf, Gtk
import cairo

from random import uniform

from sprites import Sprites, Sprite

import traceback

import logging
_logger = logging.getLogger('reflection-activity')

try:
    from sugar3.graphics import style
    GRID_CELL_SIZE = style.GRID_CELL_SIZE
except ImportError:
    GRID_CELL_SIZE = 0


LEVELS_TRUE = ['def generate_pattern(self):\n\
    # Level 1: Center dot is True\n\
    dot_list = []\n\
    for i in range(25):\n\
        dot_list.append(int(uniform(0, 2)))\n\
    dot_list[12] = 1\n\
    return dot_list\n',
               'def generate_pattern(self):\n\
    # Level 2: Corners match\n\
    dot_list = []\n\
    for i in range(25):\n\
        dot_list.append(int(uniform(0, 2)))\n\
    dot_list[4] = dot_list[0]\n\
    dot_list[20] = dot_list[0]\n\
    dot_list[24] = dot_list[0]\n\
    return dot_list\n',
               'def generate_pattern(self):\n\
    # Level 3: more than 12 True\n\
    dot_list = []\n\
    for i in range(25):\n\
        dot_list.append(int(uniform(0, 2)))\n\
    def dot_sum(dot_list):\n\
       sum = 0\n\
       for i in dot_list:\n\
           sum += i\n\
       return sum\n\
    while dot_sum(dot_list) < 13:\n\
       dot_list[int(uniform(0, 25))] = 1\n\
    return dot_list\n',
               'def generate_pattern(self):\n\
    # Level 4: even number of True\n\
    dot_list = []\n\
    for i in range(25):\n\
        dot_list.append(int(uniform(0, 2)))\n\
    def dot_sum(dot_list):\n\
       sum = 0\n\
       for i in dot_list:\n\
           sum += i\n\
       return sum\n\
    while dot_sum(dot_list) % 2 == 1:\n\
       dot_list[int(uniform(0, 25))] = 1\n\
    return dot_list\n',
               'def generate_pattern(self):\n\
    # Level 5: diagonal is True\n\
    dot_list = []\n\
    for i in range(25):\n\
        dot_list.append(int(uniform(0, 2)))\n\
    if int(uniform(0, 2)) == 1:\n\
        dot_list[0] = 1\n\
        dot_list[6] = 1\n\
        dot_list[12] = 1\n\
        dot_list[18] = 1\n\
        dot_list[24] = 1\n\
    else:\n\
        dot_list[4] = 1\n\
        dot_list[8] = 1\n\
        dot_list[12] = 1\n\
        dot_list[16] = 1\n\
        dot_list[20] = 1\n\
    return dot_list\n',
               'def generate_pattern(self):\n\
    # Level 6: True above each False\n\
    dot_list = []\n\
    for i in range(25):\n\
        dot_list.append(int(uniform(0, 2)))\n\
    for i in range(20):\n\
        j = 24 - i\n\
        if dot_list[j] == 0:\n\
            dot_list[j - 5] = 1\n\
    return dot_list\n',
               'def generate_pattern(self):\n\
    # Level 7: True in each row\n\
    dot_list = []\n\
    for i in range(25):\n\
        dot_list.append(int(uniform(0, 2)))\n\
    def row_sum(dot_list, row):\n\
       sum = 0\n\
       for i in range(5):\n\
           sum += dot_list[row * 5 + i]\n\
       return sum\n\
    for y in range(5):\n\
        if row_sum(dot_list, y) == 0:\n\
            dot_list[y * 5 + int(uniform(0, 5))] = 1\n\
    return dot_list\n',
               'def generate_pattern(self):\n\
    # Level 8: True path to center\n\
    dot_list = []\n\
    for i in range(25):\n\
        dot_list.append(int(uniform(0, 2)))\n\
    dot_list[12] = 1\n\
    paths = [[2, 7], [10, 11], [13, 14], [17, 22]]\n\
    n = int(uniform(0, 4))\n\
    dot_list[paths[n][0]] = 1\n\
    dot_list[paths[n][1]] = 1\n\
    return dot_list\n',
               'def generate_pattern(self):\n\
    # Level 9: more True on top half than on bottom half\n\
    dot_list = []\n\
    for i in range(25):\n\
        dot_list.append(int(uniform(0, 2)))\n\
    def top_sum(dot_list):\n\
       sum = 0\n\
       for i in range(10):\n\
           sum += dot_list[i]\n\
       return sum\n\
    def bot_sum(dot_list):\n\
       sum = 0\n\
       for i in range(10):\n\
           sum += dot_list[i + 15]\n\
       return sum\n\
    while top_sum(dot_list) < (bot_sum(dot_list) + 1):\n\
       dot_list[int(uniform(0, 10))] = 1\n\
    return dot_list\n',
               'def generate_pattern(self):\n\
    # Level 10: smilely face\n\
    dot_list = []\n\
    parity = int(uniform(0, 2))\n\
    for i in range(25):\n\
        dot_list.append(parity)\n\
    dot_list[6] = 1 - parity\n\
    dot_list[8] = 1 - parity\n\
    dot_list[15] = 1 - parity\n\
    dot_list[19] = 1 - parity\n\
    dot_list[21] = 1 - parity\n\
    dot_list[22] = 1 - parity\n\
    dot_list[23] = 1 - parity\n\
    return dot_list\n'
               ]
LEVELS_FALSE = ['def generate_pattern(self):\n\
    # Level 1: Center dot is False\n\
    dot_list = []\n\
    for i in range(25):\n\
        dot_list.append(int(uniform(0, 2)))\n\
    dot_list[12] = 0\n\
    return dot_list\n',
                'def generate_pattern(self):\n\
    # Level 2: Corners do not match\n\
    dot_list = []\n\
    for i in range(25):\n\
        dot_list.append(int(uniform(0, 2)))\n\
    n = int(uniform(0, 3))\n\
    corner = [4, 20, 24]\n\
    dot_list[corner[n]] = 1 - dot_list[0]\n\
    return dot_list\n',
                'def generate_pattern(self):\n\
    # Level 3: less than 13 True\n\
    dot_list = []\n\
    for i in range(25):\n\
        dot_list.append(int(uniform(0, 2)))\n\
    def dot_sum(dot_list):\n\
       sum = 0\n\
       for i in dot_list:\n\
           sum += i\n\
       return sum\n\
    while dot_sum(dot_list) > 12:\n\
       dot_list[int(uniform(0, 25))] = 0\n\
    return dot_list\n',
                'def generate_pattern(self):\n\
    # Level 4: odd number of True\n\
    dot_list = []\n\
    for i in range(25):\n\
        dot_list.append(int(uniform(0, 2)))\n\
    def dot_sum(dot_list):\n\
       sum = 0\n\
       for i in dot_list:\n\
           sum += i\n\
       return sum\n\
    while dot_sum(dot_list) % 2 == 0:\n\
       dot_list[int(uniform(0, 25))] = 1\n\
    return dot_list\n',
                'def generate_pattern(self):\n\
    # Level 5: diagonal is True\n\
    dot_list = []\n\
    for i in range(25):\n\
        dot_list.append(int(uniform(0, 2)))\n\
    if int(uniform(0, 2)) == 1:\n\
        dot_list[0] = 1\n\
        dot_list[6] = 1\n\
        dot_list[12] = 1\n\
        dot_list[18] = 1\n\
        dot_list[24] = 1\n\
        n = int(uniform(0, 5))\n\
        diagonal = [0, 6, 12, 18, 24]\n\
        dot_list[diagonal[n]] = 1 - dot_list[0]\n\
    else:\n\
        dot_list[4] = 1\n\
        dot_list[8] = 1\n\
        dot_list[12] = 1\n\
        dot_list[16] = 1\n\
        dot_list[20] = 1\n\
        n = int(uniform(0, 5))\n\
        diagonal = [4, 8, 12, 16, 20]\n\
        dot_list[diagonal[n]] = 1 - dot_list[0]\n\
    return dot_list\n',
                'def generate_pattern(self):\n\
    # Level 6: True to right of each False\n\
    dot_list = []\n\
    for i in range(25):\n\
        dot_list.append(int(uniform(0, 2)))\n\
    for y in range(5):\n\
        for x in range(4):\n\
            j = y * 5 + x\n\
            if dot_list[j] == 0:\n\
                dot_list[j + 1] = 1\n\
    return dot_list\n',
                'def generate_pattern(self):\n\
    # Level 7: One row with no True\n\
    dot_list = []\n\
    for i in range(25):\n\
        dot_list.append(int(uniform(0, 2)))\n\
    y = int(uniform(0, 5))\n\
    for i in range(5):\n\
        dot_list[y * 5 + i] = 0\n\
    return dot_list\n',
                'def generate_pattern(self):\n\
    # Level 8: No True path to center\n\
    dot_list = []\n\
    for i in range(25):\n\
        dot_list.append(int(uniform(0, 2)))\n\
    dot_list[12] = 1\n\
    paths = [[2, 7], [10, 11], [13, 14], [17, 22]]\n\
    for n in range(4):\n\
        if int(uniform(0, 2)) == 1:\n\
            dot_list[paths[n][0]] = 1\n\
            dot_list[paths[n][1]] = 0\n\
        else:\n\
            dot_list[paths[n][0]] = 0\n\
            dot_list[paths[n][1]] = 1\n\
    return dot_list\n',
                'def generate_pattern(self):\n\
    # Level 9: more True on bottom half than on top half\n\
    dot_list = []\n\
    for i in range(25):\n\
        dot_list.append(int(uniform(0, 2)))\n\
    def top_sum(dot_list):\n\
       sum = 0\n\
       for i in range(10):\n\
           sum += dot_list[i]\n\
       return sum\n\
    def bot_sum(dot_list):\n\
       sum = 0\n\
       for i in range(10):\n\
           sum += dot_list[i + 15]\n\
       return sum\n\
    while (top_sum(dot_list) + 1) > bot_sum(dot_list):\n\
       dot_list[int(uniform(0, 10)) + 15] = 1\n\
    return dot_list\n',
                'def generate_pattern(self):\n\
    # Level 10: frowny face\n\
    dot_list = []\n\
    parity = int(uniform(0, 2))\n\
    for i in range(25):\n\
        dot_list.append(parity)\n\
    dot_list[6] = 1 - parity\n\
    dot_list[8] = 1 - parity\n\
    dot_list[16] = 1 - parity\n\
    dot_list[17] = 1 - parity\n\
    dot_list[18] = 1 - parity\n\
    dot_list[20] = 1 - parity\n\
    dot_list[24] = 1 - parity\n\
    return dot_list\n']

# Grid dimensions
GRID = 5
WHITE = 2
DOT_SIZE = 80


class Game():

    def __init__(self, canvas, parent=None, colors=['#A0FFA0', '#FF8080']):
        self._activity = parent
        self._colors = [colors[0]]
        self._colors.append(colors[1])

        self._canvas = canvas
        if parent is not None:
            parent.show_all()
            self._parent = parent

        self._canvas.connect("draw", self.__draw_cb)

        self._width = Gdk.Screen.width()
        self._height = Gdk.Screen.height() - (GRID_CELL_SIZE * 1.5)
        self._scale = self._width / (10 * DOT_SIZE * 1.2)
        self._dot_size = int(DOT_SIZE * self._scale)
        self._space = int(self._dot_size / 5.)
        self.max_levels = len(LEVELS_TRUE)
        self.this_pattern = False

        # Generate the sprites we'll need...
        self._sprites = Sprites(self._canvas)
        self._dots = []
        self._generate_grid()

    def _generate_grid(self):
        ''' Make a new set of dots for a grid of size edge '''
        i = 0
        for y in range(GRID):
            for x in range(GRID):
                xoffset = int((self._width - GRID * self._dot_size -
                               (GRID - 1) * self._space) / 2.)
                if i < len(self._dots):
                    self._dots[i].move(
                        (xoffset + x * (self._dot_size + self._space),
                         y * (self._dot_size + self._space) + self._space))
                else:
                    self._dots.append(
                        Sprite(self._sprites,
                               xoffset + x * (self._dot_size + self._space),
                               y * (self._dot_size + self._space) +
                               self._space,
                               self._new_dot(self._colors[0])))
                self._dots[i].type = 0
                self._dots[-1].set_label_attributes(40)
                i += 1

    def show(self, dot_list):
        for i in range(GRID * GRID):
            self._dots[i].set_shape(self._new_dot(self._colors[dot_list[i]]))
            self._dots[i].type = dot_list[i]

    def show_true(self):
        self.show(self._generate_pattern(LEVELS_TRUE[self._activity.level]))
        self.this_pattern = True

    def show_false(self):
        self.show(self._generate_pattern(LEVELS_FALSE[self._activity.level]))
        self.this_pattern = False

    def show_random(self):
        ''' Fill the grid with a true or false pattern '''
        if int(uniform(0, 2)) == 0:
            self.show_true()
        else:
            self.show_false()

    def _initiating(self):
        return self._activity.initiating

    def new_game(self):
        ''' Start a new game. '''
        self.show_random()

    def restore_grid(self, dot_list, boolean):
        ''' Restore a grid from the share '''
        self.show(dot_list)
        self.this_pattern = boolean

    def save_grid(self):
        ''' Return dot list for sharing '''
        dot_list = []
        for dot in self._dots:
            dot_list.append(dot.type)
        return(dot_list, self.this_pattern)

    def _grid_to_dot(self, pos):
        ''' calculate the dot index from a column and row in the grid '''
        return pos[0] + pos[1] * GRID

    def _dot_to_grid(self, dot):
        ''' calculate the grid column and row for a dot '''
        return [dot % GRID, int(dot // GRID)]

    def _set_label(self, string):
        ''' Set the label in the toolbar or the window frame. '''
        self._activity.status.set_label(string)

    def _generate_pattern(self, f):
        ''' Run Python code passed as argument '''
        userdefined = {}
        try:
            exec(f, globals(), userdefined)
            return userdefined['generate_pattern'](self)
        except ZeroDivisionError as e:
            self._set_label('Python zero-divide error: %s' % (str(e)))
        except ValueError as e:
            self._set_label('Python value error: %s' % (str(e)))
        except SyntaxError as e:
            self._set_label('Python syntax error: %s' % (str(e)))
        except NameError as e:
            self._set_label('Python name error: %s' % (str(e)))
        except OverflowError as e:
            self._set_label('Python overflow error: %s' % (str(e)))
        except TypeError as e:
            self._set_label('Python type error: %s' % (str(e)))
        except BaseException:
            self._set_label('Python error')
        traceback.print_exc()
        return None

    def __draw_cb(self, canvas, cr):
        self._sprites.redraw_sprites(cr=cr)

    def do_expose_event(self, event):
        ''' Handle the expose-event by drawing '''
        # Restrict Cairo to the exposed area
        cr = self._canvas.window.cairo_create()
        cr.rectangle(event.area.x, event.area.y,
                     event.area.width, event.area.height)
        cr.clip()
        # Refresh sprite list
        self._sprites.redraw_sprites(cr=cr)

    def _destroy_cb(self, win, event):
        Gtk.main_quit()

    def _new_dot(self, color):
        ''' generate a dot of a color color '''
        self._dot_cache = {}
        if color not in self._dot_cache:
            self._stroke = color
            self._fill = color
            self._svg_width = self._dot_size
            self._svg_height = self._dot_size
            pixbuf = svg_str_to_pixbuf(
                self._header() +
                self._circle(self._dot_size / 2., self._dot_size / 2.,
                             self._dot_size / 2.) +
                self._footer())

            surface = cairo.ImageSurface(cairo.FORMAT_ARGB32,
                                         self._svg_width, self._svg_height)
            context = cairo.Context(surface)
            Gdk.cairo_set_source_pixbuf(context, pixbuf, 0, 0)
            context.rectangle(0, 0, self._svg_width, self._svg_height)
            context.fill()
            self._dot_cache[color] = surface

        return self._dot_cache[color]

    def _header(self):
        return '<svg\n' + 'xmlns:svg="http://www.w3.org/2000/svg"\n' + \
            'xmlns="http://www.w3.org/2000/svg"\n' + \
            'xmlns:xlink="http://www.w3.org/1999/xlink"\n' + \
            'version="1.1"\n' + 'width="' + str(self._svg_width) + '"\n' + \
            'height="' + str(self._svg_height) + '">\n'

    def _circle(self, r, cx, cy):
        return '<circle style="fill:' + str(self._fill) + ';stroke:' + \
            str(self._stroke) + ';" r="' + str(r - 0.5) + '" cx="' + \
            str(cx) + '" cy="' + str(cy) + '" />\n'

    def _footer(self):
        return '</svg>\n'


def svg_str_to_pixbuf(svg_string):
    """ Load pixbuf from SVG string """
    pl = GdkPixbuf.PixbufLoader.new_with_type('svg')
    pl.write(svg_string.encode('utf-8'))
    pl.close()
    pixbuf = pl.get_pixbuf()
    return pixbuf


DeductoActivity.py
# -*- coding: utf-8 -*-
# Copyright (c) 2012 Walter Bender
# Ported to gtk3:
# Ignacio Rodríguez <ignaciorodriguez@sugarlabs.org>

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# You should have received a copy of the GNU General Public License
# along with this library; if not, write to the Free Software
# Foundation, 51 Franklin Street, Suite 500 Boston, MA 02110-1335 USA

import gi
gi.require_version('Gtk', '3.0')
gi.require_version('TelepathyGLib', '0.12')

from gi.repository import Gtk
from gi.repository import Gdk
from gi.repository import TelepathyGLib

from sugar3.activity import activity
from sugar3 import profile
from sugar3.graphics.toolbarbox import ToolbarBox
from sugar3.activity.widgets import ActivityToolbarButton
from sugar3.activity.widgets import StopButton
from sugar3.graphics.objectchooser import ObjectChooser

from toolbar_utils import button_factory, label_factory, separator_factory
from utils import json_load, json_dump

import dbus
from dbus.service import signal
from dbus.gi_service import ExportedGObject
from sugar3.presence import presenceservice
from sugar3.presence.tubeconn import TubeConnection

from gettext import gettext as _

from game import Game, LEVELS_TRUE, LEVELS_FALSE

import logging
_logger = logging.getLogger('deducto-activity')


SERVICE = 'in.seeta.Deducto'
IFACE = SERVICE
PATH = '/in/seeta/DeductoActivity'



class DeductoActivity(activity.Activity):
    """ Logic puzzle game """

    def __init__(self, handle):
        """ Initialize the toolbars and the game board """
        try:
            super(DeductoActivity, self).__init__(handle)
        except dbus.exceptions.DBusException as e:
            _logger.error(str(e))

        self.nick = profile.get_nick_name()
        if profile.get_color() is not None:
            self.colors = profile.get_color().to_string().split(',')
        else:
            self.colors = ['#A0FFA0', '#FF8080']

        self.level = 0
        self._correct = 0
        self._playing = True
        self._game_over = False

        self._python_code = None

        self._setup_toolbars()
        self._setup_dispatch_table()

        # Create a canvas
        canvas = Gtk.DrawingArea()
        canvas.set_size_request(Gdk.Screen.width(),
                                Gdk.Screen.height())
        self.set_canvas(canvas)
        canvas.show()
        self.show_all()

        self._game = Game(canvas, parent=self, colors=self.colors)

        self._sharing = False
        self._initiating = False
        self._setup_presence_service()

        if 'level' in self.metadata:
            self.level = int(self.metadata['level'])
            self.status.set_label(_('Resuming level %d') % (self.level + 1))
            self._game.show_random()
        else:
            self._game.new_game()

    def _setup_toolbars(self):
        """ Setup the toolbars. """

        self.max_participants = 4

        toolbox = ToolbarBox()

        # Activity toolbar
        activity_button = ActivityToolbarButton(self)

        toolbox.toolbar.insert(activity_button, 0)
        activity_button.show()

        self.set_toolbar_box(toolbox)
        toolbox.show()
        self.toolbar = toolbox.toolbar

        self._new_game_button = button_factory(
            'new-game', self.toolbar, self._new_game_cb,
            tooltip=_('Start a new game.'))

        separator_factory(toolbox.toolbar, False, True)

        self._true_button = button_factory(
            'true', self.toolbar, self._true_cb,
            tooltip=_('The pattern matches the rule.'))

        self._false_button = button_factory(
            'false', self.toolbar, self._false_cb,
            tooltip=_('The pattern does not match the rule.'))

        separator_factory(toolbox.toolbar, False, True)

        self._example_button = button_factory(
            'example', self.toolbar, self._example_cb,
            tooltip=_('Explore some examples.'))

        self.status = label_factory(self.toolbar, '', width=300)

        separator_factory(toolbox.toolbar, True, False)

        self._gear_button = button_factory(
            'view-source', self.toolbar,
            self._gear_cb,
            tooltip=_('Load a custom level.'))

        stop_button = StopButton(self)
        stop_button.props.accelerator = '<Ctrl>q'
        toolbox.toolbar.insert(stop_button, -1)
        stop_button.show()

    def _new_game_cb(self, button=None):
        ''' Start a new game. '''
        if (not self._sharing) or self._initiating:
            self._game_over = False
            self._correct = 0
            self.level = 0
            if not self._playing:
                self._example_cb()
            self._game.new_game()
            if self._initiating:
                _logger.debug('sending new game and new grid')
                self._send_new_game()
                self._send_new_grid()
            self.status.set_label(_('Playing level %d') % (self.level + 1))
        else:
            self.status.set_label(_('Only sharer can start a new game.'))

    def _test_for_game_over(self):
        ''' If we are at maximum levels, the game is over '''
        if self.level == self._game.max_levels:
            self.level = 0
            self._game_over = True
            self.status.set_label(_('Game over.'))
        else:
            self.status.set_label(_('Playing level %d') % (self.level + 1))
            self._correct = 0
            if (not self._sharing) or self._initiating:
                self._game.show_random()
                if self._initiating:
                    self._send_new_grid()

    def _true_cb(self, button=None):
        ''' Declare pattern true or show an example of a true pattern. '''
        if self._game_over:
            if (not self._sharing) or self._initiating:
                self.status.set_label(_('Click on new game button to begin.'))
            else:
                self.status.set_label(_('Wait for sharer to start ' +
                                        'a new game.'))
            return
        if self._playing:
            if self._game.this_pattern:
                self._correct += 1
                if self._correct == 5:
                    self.level += 1
                    self._test_for_game_over()
                    self.metadata['level'] = str(self.level)
                else:
                    self.status.set_label(
                        _('%d correct answers.') % (self._correct))
                    if (not self._sharing) or self._initiating:
                        self._game.show_random()
                        if self._initiating:
                            self._send_new_grid()
            else:
                self.status.set_label(_('Pattern was false.'))
                self._correct = 0
            if (button is not None) and self._sharing:
                self._send_true_button_click()
        else:
            self._game.show_true()

    def _false_cb(self, button=None):
        ''' Declare pattern false or show an example of a false pattern. '''
        if self._game_over:
            if (not self._sharing) or self._initiating:
                self.status.set_label(_('Click on new game button to begin.'))
            else:
                self.status.set_label(_('Wait for sharer to start ' +
                                        'a new game.'))
            return
        if self._playing:
            if not self._game.this_pattern:
                self._correct += 1
                if self._correct == 5:
                    self.level += 1
                    self._test_for_game_over()
                else:
                    self.status.set_label(
                        _('%d correct answers.') % (self._correct))
                    if (not self._sharing) or self._initiating:
                        self._game.show_random()
                        if self._initiating:
                            self._send_new_grid()
            else:
                self.status.set_label(_('Pattern was true.'))
                self._correct = 0
            if (button is not None) and self._sharing:
                self._send_false_button_click()
        else:
            self._game.show_false()

    def _example_cb(self, button=None):
        ''' Show examples or resume play of current level. '''
        if self._playing:
            self._example_button.set_icon_name('resume-play')
            self._example_button.set_tooltip(_('Resume play'))
            self._true_button.set_tooltip(
                _('Show a pattern that matches the rule.'))
            self._false_button.set_tooltip(
                _('Show a pattern that does not match the rule.'))
            Button1 = '☑'
            Button2 = '☒'
            self.status.set_label(_("Explore patterns with the {} and {} buttons.".format(Button1, Button2)))
            self._playing = False
        else:
            self._example_button.set_icon_name('example')
            self._example_button.set_tooltip(_('Explore some examples.'))
            self._true_button.set_tooltip(
                _('The pattern matches the rule.'))
            self._false_button.set_tooltip(
                _('The pattern does not match the rule.'))
            self.status.set_label(_('Playing level %d') % (self.level + 1))
            self._playing = True
            self._correct = 0

    def _gear_cb(self, button=None):
        ''' Load a custom level. '''
        self.status.set_text(
            _('Load a "True" pattern generator from the journal'))
        self._chooser('org.laptop.Pippy',
                      self._load_python_code_from_journal)
        if self._python_code is None:
            return
        LEVELS_TRUE.append(self._python_code)
        self.status.set_text(
            _('Load a "False" pattern generator from the journal'))
        self._chooser('org.laptop.Pippy',
                      self._load_python_code_from_journal)
        LEVELS_FALSE.append(self._python_code)
        if self._python_code is None:
            return
        self.status.set_text(_('New level added'))
        self._game.max_levels += 1

    def _load_python_code_from_journal(self, dsobject):
        ''' Read the Python code from the Journal object '''
        self._python_code = None
        try:
            _logger.debug("opening %s " % dsobject.file_path)
            file_handle = open(dsobject.file_path, "r")
            self._python_code = file_handle.read()
            file_handle.close()
        except IOError:
            _logger.debug("couldn't open %s" % dsobject.file_path)

    def _chooser(self, filter, action):
        ''' Choose an object from the datastore and take some action '''
        chooser = None
        try:
            chooser = ObjectChooser(parent=self, what_filter=filter)
        except TypeError:
            chooser = ObjectChooser(
                None, self,
                Gtk.DialogFlags.MODAL | Gtk.DialogFlags.DESTROY_WITH_PARENT)
        if chooser is not None:
            try:
                result = chooser.run()
                if result == Gtk.ResponseType.ACCEPT:
                    dsobject = chooser.get_selected_object()
                    action(dsobject)
                    dsobject.destroy()
            finally:
                chooser.destroy()
                del chooser

    # Collaboration-related methods

    # The sharer sends patterns and everyone shares whatever vote is
    # cast first among all the sharer and joiners.

    def _setup_presence_service(self):
        ''' Setup the Presence Service. '''
        self.pservice = presenceservice.get_instance()
        self._initiating = None  # sharing (True) or joining (False)

        owner = self.pservice.get_owner()
        self.owner = owner
        self._share = ""
        self.connect('shared', self._shared_cb)
        self.connect('joined', self._joined_cb)

    def _shared_cb(self, activity):
        ''' Either set up initial share...'''
        self._new_tube_common(True)

    def _joined_cb(self, activity):
        ''' ...or join an exisiting share. '''
        self._new_tube_common(False)

    def _new_tube_common(self, sharer):
        ''' Joining and sharing are mostly the same... '''
        if self._shared_activity is None:
            _logger.debug("Error: Failed to share or join activity ... \
                _shared_activity is null in _shared_cb()")
            return

        self._initiating = sharer
        self._sharing = True

        self.conn = self._shared_activity.telepathy_conn
        self.tubes_chan = self._shared_activity.telepathy_tubes_chan
        self.text_chan = self._shared_activity.telepathy_text_chan

        self.tubes_chan[TelepathyGLib.IFACE_CHANNEL_TYPE_TUBES].connect_to_signal(
            'NewTube', self._new_tube_cb)

        if sharer:
            _logger.debug('This is my activity: making a tube...')
        else:
            _logger.debug('I am joining an activity: waiting for a tube...')
            self.tubes_chan[TelepathyGLib.IFACE_CHANNEL_TYPE_TUBES].ListTubes(
                reply_handler=self._list_tubes_reply_cb,
                error_handler=self._list_tubes_error_cb)

    def _list_tubes_reply_cb(self, tubes):
        ''' Reply to a list request. '''
        for tube_info in tubes:
            self._new_tube_cb(*tube_info)

    def _list_tubes_error_cb(self, e):
        ''' Log errors. '''
        _logger.debug('Error: ListTubes() failed: %s' % (e))

    def _new_tube_cb(self, id, initiator, type, service, params, state):
        ''' Create a new tube. '''
        _logger.debug('New tube: ID=%d initator=%d type=%d service=%s \
params=%r state=%d' % (id, initiator, type, service, params, state))

        if (type == TelepathyGLib.TubeType.DBUS and service == SERVICE):
            if state == TelepathyGLib.TubeState.LOCAL_PENDING:
                self.tubes_chan[
                    TelepathyGLib.IFACE_CHANNEL_TYPE_TUBES].AcceptDBusTube(id)

            tube_conn = TubeConnection(self.conn,
                                       self.tubes_chan[TelepathyGLib.IFACE_CHANNEL_TYPE_TUBES],
                                       id,
                                       group_iface=self. text_chan[TelepathyGLib.IFACE_CHANNEL_INTERFACE_GROUP])

            self.chattube = ChatTube(tube_conn, self._initiating,
                                     self._event_received_cb)

    def _setup_dispatch_table(self):
        ''' Associate tokens with commands. '''
        self._processing_methods = {
            'n': [self._receive_new_game, 'new game'],
            'g': [self._receive_new_grid, 'get a new grid'],
            't': [self._receive_true_button_click, 'get a true button press'],
            'f': [self._receive_false_button_click, 'get a false ' +
                                                    'button press'],
        }

    def _event_received_cb(self, event_message):
        ''' Data from a tube has arrived. '''
        if len(event_message) == 0:
            return
        try:
            command, payload = event_message.split('|', 2)
        except ValueError:
            _logger.debug('Could not split event message %s' % (event_message))
            return
        self._processing_methods[command][0](payload)

    def _send_new_game(self):
        '''
        Send a new game message to all players
        (only sharer sends grids)
        '''
        self._send_event('n| ')

    def _receive_new_game(self, payload):
        ''' Receive a new game notification from the sharer. '''
        self._game_over = False
        self._correct = 0
        self.level = 0
        if not self._playing:
            self._example_cb()
        self.status.set_label(_('Playing level %d') % (self.level + 1))

    def _send_new_grid(self):
        ''' Send a new grid to all players (only sharer sends grids) '''
        self._send_event('g|%s' % (json_dump(self._game.save_grid())))

    def _receive_new_grid(self, payload):
        ''' Receive a grid from the sharer. '''
        (dot_list, boolean) = json_load(payload)
        self._game.restore_grid(dot_list, boolean)

    def _send_true_button_click(self):
        ''' Send a true click to all the players '''
        self._send_event('t|t')

    def _receive_true_button_click(self, payload):
        ''' When a button is clicked, everyone should react. '''
        self._playing = True
        self._true_cb()

    def _send_false_button_click(self):
        ''' Send a false click to all the players '''
        self._send_event('f|f')

    def _receive_false_button_click(self, payload):
        ''' When a button is clicked, everyone should react. '''
        self._playing = True
        self._false_cb()

    def _send_event(self, entry):
        ''' Send event through the tube. '''
        if hasattr(self, 'chattube') and self.chattube is not None:
            self.chattube.SendText(entry)


class ChatTube(ExportedGObject):
    ''' Class for setting up tube for sharing '''

    def __init__(self, tube, is_initiator, stack_received_cb):
        super(ChatTube, self).__init__(tube, PATH)
        self.tube = tube
        self.is_initiator = is_initiator  # Are we sharing or joining activity?
        self.stack_received_cb = stack_received_cb
        self.stack = ''

        self.tube.add_signal_receiver(self._send_stack_cb, 'SendText', IFACE,
                                      path=PATH, sender_keyword='sender')

    def _send_stack_cb(self, text, sender=None):
        if sender == self.tube.get_unique_name():
            return
        self.stack = text
        self.stack_received_cb(text)

    @signal(dbus_interface=IFACE, signature='s')
    def SendText(self, text):
        self.stack = text
